src/libraries/System.Private.CoreLib/src/System/IO/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -0,0 +1,406 @@
+            }
+            catch (ArgumentException ex)
+            {
+                throw new IOException(SR.IO_BindHandleFailed, ex);
+            }
+            finally
+            {
+                if (_fileHandle.ThreadPoolBinding == null)
+                {
+                    // We should close the handle so that the handle is not open until SafeFileHandle GC
+                    Debug.Assert(!_exposedHandle, "Are we closing handle that we exposed/not own, how?");
+                    _fileHandle.Dispose();
+                }
+            }
+        }
+
+        // called by BufferedStream
+        internal void OnBufferAllocated(byte[] buffer)
+        {
+            Debug.Assert(buffer != null);
+            Debug.Assert(_preallocatedOverlapped == null);
+
+            _preallocatedOverlapped = new PreAllocatedOverlapped(FileStreamCompletionSource.s_ioCallback, this, buffer);
+        }
+
+        SafeFileHandle IFileStreamCompletionSourceStrategy.FileHandle => _fileHandle;
+
+        FileStreamCompletionSource? IFileStreamCompletionSourceStrategy.CurrentOverlappedOwner => _currentOverlappedOwner;
+
+        FileStreamCompletionSource? IFileStreamCompletionSourceStrategy.CompareExchangeCurrentOverlappedOwner(FileStreamCompletionSource? newSource, FileStreamCompletionSource? existingSource)
+            => Interlocked.CompareExchange(ref _currentOverlappedOwner, newSource, existingSource);
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => ReadAsyncInternal(new Memory<byte>(buffer, offset, count)).GetAwaiter().GetResult();
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            => ReadAsyncInternal(new Memory<byte>(buffer, offset, count), cancellationToken);
+
+        public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken = default)
+            => new ValueTask<int>(ReadAsyncInternal(destination, cancellationToken));
+
+        private unsafe Task<int> ReadAsyncInternal(Memory<byte> destination, CancellationToken cancellationToken = default)
+        {
+            if (!CanRead)
+            {
+                throw Error.GetReadNotSupported();
+            }
+
+            Debug.Assert(!_fileHandle.IsClosed, "!_handle.IsClosed");
+
+            // Create and store async stream class library specific data in the async result
+            FileStreamCompletionSource completionSource = FileStreamCompletionSource.Create(this, _preallocatedOverlapped, 0, destination);
+            NativeOverlapped* intOverlapped = completionSource.Overlapped;
+
+            // Calculate position in the file we should be at after the read is done
+            if (CanSeek)
+            {
+                long len = Length;
+
+                // Make sure we are reading from the position that we think we are
+                VerifyOSHandlePosition();
```

Comment:
We're going to remove these subsequently, right?  Just making sure :)

Reply:
Yes, I already have a PR for that (but it's blocked by this PR): https://github.com/dotnet/runtime/pull/49145 

Changes: https://github.com/dotnet/runtime/commit/d06f1752be351b7d5a48e2a4de79167224fe66b4